### PR TITLE
AV-166743: Enable ICAP Scanning through HostRule

### PR DESCRIPTION
**Pull request by saurabhvaidvmware**

### CODE REVIEW

**Comment by saurabhvaidvmware** [comment]
build ako

File: internal/nodes/avi_model_nodes.go [hunk 1]
```diff
@@ -374,6 +374,7 @@ type AviVsNode struct {
 	Tenant                string
 	ServiceEngineGroup    string
 	ApplicationProfile    string
+	ICAPProfile           string
```
Comment (swathinsankaran) [inline internal/nodes/avi_model_nodes.go:377]: Is this field being used somewhere? I am seeing two fields with similar intents.

File: internal/rest/avi_evh_obj.go [hunk 2]
```diff
@@ -492,6 +501,10 @@ func (rest *RestOperations) AviVsChildEvhBuild(vs_meta *nodes.AviEvhVsNode, rest
 		VhType:                proto.String(utils.VS_TYPE_VH_ENHANCED),
 	}
 
+	if len(vs_meta.ICAPProfileRefs) != 0 {
+		evhChild.IcapRequestProfileRefs = &vs_meta.ICAPProfileRefs
```
Comment (swathinsankaran) [inline internal/rest/avi_evh_obj.go:505]: Do we have to attach the ICAP profiles to SNI Child VS also?

File: vendor/github.com/vmware/alb-sdk/go/models/virtual_service.go [hunk 3]
```diff
@@ -113,7 +113,7 @@ type VirtualService struct {
 	HTTPPolicies []*HTTPPolicies `json:"http_policies,omitempty"`
 
 	// The config settings for the ICAP server when checking the HTTP request. It is a reference to an object of type IcapProfile. Field introduced in 20.1.1. Maximum of 1 items allowed. Allowed in Enterprise edition with any value, Enterprise with Cloud Services edition.
-	IcapRequestProfileRefs []string `json:"icap_request_profile_refs,omitempty"`
+	IcapRequestProfileRefs *[]string `json:"icap_request_profile_refs,omitempty"`
```
Comment (swathinsankaran) [inline vendor/github.com/vmware/alb-sdk/go/models/virtual_service.go:116]: Is this change done by mistake? Ideally, we won't make any changes to files in the vendor folders.

**Comment by swathinsankaran** [comment]
@saurabhvaidvmware Could you please add UTs or will it be taken care in a separate PR?

File: internal/nodes/avi_model_nodes.go (hunk 1 above)
Comment (saurabhvaidvmware) [reply internal/nodes/avi_model_nodes.go:377]: I had been using it to build the reference string, but then dropped that. I'll remove this entry.

File: vendor/github.com/vmware/alb-sdk/go/models/virtual_service.go (hunk 3 above)
Comment (saurabhvaidvmware) [reply vendor/github.com/vmware/alb-sdk/go/models/virtual_service.go:116]: I wasn't aware of that, will revert it back and change all references accordingly.

**Comment by saurabhvaidvmware** [comment]
> @saurabhvaidvmware Could you please add UTs or will it be taken care of in a separate PR?

@swathinsankaran I'm currently working on the UT's and will push the changes in this PR only shortly

File: internal/rest/avi_evh_obj.go (hunk 2 above)
Comment (saurabhvaidvmware) [reply internal/rest/avi_evh_obj.go]: Working on implementing ICAP for SNI mode also

Comment (saurabhvaidvmware) [reply internal/rest/avi_evh_obj.go]: Have added code for adding ICAP profile for SNI child VS

File: tests/ingresstests/crd_test.go [hunk 4]
```diff
@@ -54,6 +54,8 @@ func TestCreateDeleteHostRule(t *testing.T) {
 	g.Expect(nodes[0].SniNodes[0].SSLKeyCertAviRef[0]).To(gomega.ContainSubstring("thisisaviref-sslkey"))
 	g.Expect(nodes[0].SniNodes[0].WafPolicyRef).To(gomega.ContainSubstring("thisisaviref-waf"))
 	g.Expect(nodes[0].SniNodes[0].AppProfileRef).To(gomega.ContainSubstring("thisisaviref-appprof"))
+	g.Expect(nodes[0].SniNodes[0].ICAPProfileRefs).To(gomega.HaveLen(1))
```
Comment (swathinsankaran) [inline tests/ingresstests/crd_test.go:57]: @saurabhvaidvmware Please add similar UTs for dedicated VS also.

**Review by swathinsankaran — APPROVED** [review]
LGTM.

**Review by arihantg — APPROVED** [review]

**Review by swathinsankaran — APPROVED** [review]

**Comment by alex1989hu** [comment]
Thank you @saurabhvaidvmware @swathinsankaran

**Pull request merged**